### DISCUSSION TITLE
Wrapping transformer models for text classification

### DISCUSSION THREAD

**Discussion by adrianeboyd** [explosion/spaCy#10200] [Help: Coding & Implementations]
Questions related to the development of [`spacy-wrap`](https://github.com/KennethEnevoldsen/spacy-wrap/), which wraps transformer text classification models for inference:

- Haven't actually tested whether the model would be able to train. Ideally, the user should be able to further fine-tune the model. I assume setting the `update` and `predict` extensions would solve this issue. I have created an [issue](https://github.com/KennethEnevoldsen/spacy-wrap/issues/4) for it
- Will make sure to add a license
- Regarding the maintenance, this is actually a reworked version of a version in DaCy which broke with the spacy-transformers update, however, what I understand from your suggestion is that I copy-paste over the code to avoid breaking changes with changes in spacy-transformers? Or am I misunderstanding?
- If you have the time while developing this I got a few additional questions about whether it is possible to set the [assigns](https://github.com/explosion/spacy-transformers/blob/5a36943fccb66b5e7c7c2079b1b90ff9b2f9d020/spacy_transformers/pipeline_component.py#L49) dynamically? Similarly, I was also wondering what the [dim "n0"](https://github.com/explosion/spacy-transformers/blob/5a36943fccb66b5e7c7c2079b1b90ff9b2f9d020/spacy_transformers/layers/transformer_model.py#L158) is used for?

_Originally posted by @KennethEnevoldsen in https://github.com/explosion/spaCy/issues/10168#issuecomment-1028934471_

**Comment by adrianeboyd** [comment]
Well, that wasn't 100% what I was going for in the conversion, but here we have a discussion at least, since I think this may be interesting for others interested in using similar models.

* The `transformer` component is only updated if other components like a tagger are listening to it. If you want to make it trainable like a typical annotating pipeline component, there's a lot of machinery in `Transformer` that's not really set up the right way.

  It might be easier not to extend `Transformer` but instead extend `TrainablePipe` and set up the model as an inline transformer in the model architecture (something like `spacy-transformers.TransformerModel.v3`) and then implement `predict`, `set_annotations`, and `get_loss` to handle the extensions and training. The inline transformer model serialization should just work (we hope, as of `spacy-transformers` v1.1, it has been a total pain to implement). I think your component would look a lot more like `textcat` and less like `transformer` in the end.

  (And this is a side issue, but it was a bad choice to have the transformer config settings in the `[components]` block. They really should be in the `[initialize]` block, but we don't want to change the design for `transformer` at this point.)

* Currently there isn't any templating in the component factory definitions, which is an issue we're aware of, e.g., for the scores for `spancat` which actually depend on the `spans_key` setting. This would be a pretty non-trivial extension (I think, anyway, maybe I'm overestimating it) and it's not something that we're currently working on in any way.

  There are also some components where `assigns` could potentially be determined dynamically, but a single static factory definition isn't really possible (like for the `attribute_ruler`), so the static pipe analysis tools don't really work like you would want.

* The listener details are not my specialty, but I think that `nO` is used in turn to initialize the dimensions for the `TransformerListener` that's listening to the transformer in another component.

* It's kind of unclear which code exactly is copied and pasted (also from a licensing standpoint).

Edited to try to straighten out the listener terminology.

**Reply by svlandeg** [reply]
> The listener details are not my specialty, but I think that `nO` is used in turn to initialize the dimensions for the `TransformerListener` that's listening to the transformer in another component.

It might be helpful to link the [developer docs on the listeners](https://github.com/explosion/spaCy/blob/master/extra/DEVELOPER_DOCS/Listeners.md), specifically section 2B about "Shape inference", that provides some more information about the `nO` output dimension setting and why it's set the way it is.

**Reply by KennethEnevoldsen** [reply]
Thanks for this feedback,

- I will try to restructure it using the `TrainablePipe` instead. 

- Thanks for the clarification I will just leave it unspecified for now. Then if the user wish to wrap a specific model then can set the assigns for that specifically.

- Seems like I don't need the `n0`, will take a look at the docs regardless.

- I will clear up what code is copied and what isn't. Would a docstring in the start of the .py files be preferable?

**Reply by adrianeboyd** [reply]
It seems like it ought to be really simple to extend from `spacy-transformers`, but switching from `AutoModel` to `AutoModelForSequenceClassification` looks like it's a huge headache overall. What I said above about serialization isn't true for this use, and I don't think you can save and reload `spacy-wrap` pipelines at this point?

In terms of the code, I was mainly expecting to be able to find the copyright notice and a copy of the license from `spacy-transformers` at least once in some form.

**Reply by KennethEnevoldsen** [reply]
Started working on this now

- No the serialization does not seem to work:

<details>
  <summary>Traceback</summary>


```
RuntimeError                              Traceback (most recent call last)
/Users/au561649/Desktop/Github/spacy-wrap/tmp.py in <module>
----> [43]()[ transformer.from_disk("test")

File ~/.virtualenvs/spacy_wrap/lib/python3.8/site-packages/spacy_transformers/pipeline_component.py:420, in Transformer.from_disk(self, path, exclude)
    ]()[413]()[         self.model.attrs["set_transformer"](self.model, hf_model)
    ]()[415]()[ deserialize = {
    ]()[416]()[     "vocab": self.vocab.from_disk,
    ]()[417]()[     "cfg": lambda p: self.cfg.update(deserialize_config(p)),
    ]()[418]()[     "model": load_model,
    ]()[419]()[ }
--> ]()[420]()[ util.from_disk(path, deserialize, exclude)
    ]()[421]()[ return self

File ~/.virtualenvs/spacy_wrap/lib/python3.8/site-packages/spacy/util.py:1300, in from_disk(path, readers, exclude)
   ]()[1297]()[ for key, reader in readers.items():
   ]()[1298]()[     # Split to support file names like meta.json
   ]()[1299]()[     if key.split(".")[0] not in exclude:
-> ]()[1300]()[         reader(path / key)
   ]()[1301]()[ return path

File ~/.virtualenvs/spacy_wrap/lib/python3.8/site-packages/spacy_transformers/pipeline_component.py:394, in Transformer.from_disk.<locals>.load_model(p)
    ]()[392]()[ try:
    ]()[393]()[     with open(p, "rb") as mfile:
--> ]()[394]()[         self.model.from_bytes(mfile.read())
    ]()[395]()[ except AttributeError:
    ]()[396]()[     raise ValueError(Errors.E149) from None

File ~/.virtualenvs/spacy_wrap/lib/python3.8/site-packages/thinc/model.py:593, in Model.from_bytes(self, bytes_data)
    ]()[591]()[ msg = srsly.msgpack_loads(bytes_data)
    ]()[592]()[ msg = convert_recursive(is_xp_array, self.ops.asarray, msg)
--> ]()[593]()[ return self.from_dict(msg)

File ~/.virtualenvs/spacy_wrap/lib/python3.8/site-packages/thinc/model.py:631, in Model.from_dict(self, msg)
    ]()[629]()[         node.set_param(param_name, value)
    ]()[630]()[     for i, shim_bytes in enumerate(msg["shims"][i]):
--> ]()[631]()[         node.shims[i].from_bytes(shim_bytes)
    ]()[632]()[ return self

File ~/.virtualenvs/spacy_wrap/lib/python3.8/site-packages/spacy_transformers/layers/hf_shim.py:117, in HFShim.from_bytes(self, bytes_data)
    ]()[115]()[         device_id = torch.cuda.current_device()
    ]()[116]()[         map_location = f"cuda:{device_id}"
--> ]()[117]()[     self._model.load_state_dict(torch.load(filelike, map_location=map_location))
    ]()[118]()[     self._model.to(map_location)
    ]()[119]()[ else:

File ~/.virtualenvs/spacy_wrap/lib/python3.8/site-packages/torch/nn/modules/module.py:1482, in Module.load_state_dict(self, state_dict, strict)
   ]()[1477]()[         error_msgs.insert(
   ]()[1478]()[             0, 'Missing key(s) in state_dict: {}. '.format(
   ]()[1479]()[                 ', '.join('"{}"'.format(k) for k in missing_keys)))
   ]()[1481]()[ if len(error_msgs) > 0:
-> ]()[1482]()[     raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
   ]()[1483]()[                        self.__class__.__name__, "\n\t".join(error_msgs)))
   ]()[1484]()[ return _IncompatibleKeys(missing_keys, unexpected_keys)

RuntimeError: Error(s) in loading state_dict for RobertaModel:]()
```

</details>

- I have now added the MIT license to spacy-wrap as well, should I place the copyright notice at the docstring of the relevant scripts?

**Reply by KennethEnevoldsen** [reply]
Just wanted to note that I have added the copyright notices along with code changes in the docstring of the relevant documents:
https://github.com/KennethEnevoldsen/spacy-wrap/tree/Refactor/clf_trf
(let me know if there is something I should change)

Tried to play around with the serialization as well. Seems like the problem resides in the keys wanting to be loaded using: 
```
Missing key(s) in state_dict: "embeddings.position_ids", [...] "encoder.layer.0.attention.output.dense.bias", [...]
```

but being loaded as:
```
Unexpected key(s) in state_dict: "roberta.embeddings.position_ids", [...] "roberta.encoder.layer.0.attention.output.dense.bias", [...]
```
*notice the 'roberta.'-prefix*

I am not entirely sure where this goes wrong though. 
I will probably start working on setting it up such that the model should be trainable, but might get back to the serialization.

**Reply by adrianeboyd** [reply]
I think the problem is that it's being reloaded with `AutoModel`. There are two places where you have `AutoModel`, one in `huggingface_from_pretrained` that used on initialization, and one in `HFShim` that's used for internal serialization. I'm not sure how to make this work better. There's not an option you can provide to `AutoModel` instead of using `AutoModelForSequenceClassification`, is there?

(You are also an author who should included as a copyright holder, too!)

**Reply by KennethEnevoldsen** [reply]
Ahh, I will take a look at it!

**Reply by KennethEnevoldsen** [reply]
Hi @adrianeboyd, long time since I last looked at this, but have gotten some time so wanted to take a look at it again?

Seems like it is not possible to use the `AutoModel` approach (at least I far as I could figure out). Might it be possible to add an option to change the loaders using a syntax like (`HFShim`):

```
    def __init__(
        self,
        model: HFObjects,
        config=None,
        optimizer: Any = None,
        mixed_precision: bool = False,
        grad_scaler_config: dict = {},
        load_config_fn: Callable = AutoConfig.from_pretrained,
        load_tokenizer_fn: Callable = AutoTokenizer.from_pretrained,
        load_model_from_config_fn: Callable = AutoModel.from_config
    )
```

It should be backward compatible and allow passing these arguments forward in the `HFWrapper`.

```
def HFWrapper(
    hf_model: "HFObjects",
    convert_inputs: Optional[Callable] = None,
    convert_outputs: Optional[Callable] = None,
    mixed_precision: bool = False,
    grad_scaler_config: dict = {},
    load_config_fn: Callable = AutoConfig.from_pretrained,
    load_tokenizer_fn: Callable = AutoTokenizer.from_pretrained,
    load_model_from_config_fn: Callable = AutoModel.from_config
)
```

if so I can create a PR on this. My alternative solution currently is to use a monkey patch or define two classes called `HFClassificationWrapper` and an `HFClassificationShim`.

**Reply by adrianeboyd** [reply]
That is an interesting suggestion that I hadn't thought of. Is it straightforward to pass the settings to `HFShim`?

**Reply by KennethEnevoldsen** [reply]
That is at least what I have done as the temporary solution for spacy-wrap for the [HFWrapper](https://github.com/KennethEnevoldsen/spacy-wrap/blob/main/spacy_wrap/layers/hf_classification_wrapper.py) and [HFShim](https://github.com/KennethEnevoldsen/spacy-wrap/blob/main/spacy_wrap/layers/hf_shim.py). Which is called [here](https://github.com/KennethEnevoldsen/spacy-wrap/blob/96d333b7f7dc4157b365de477b158d64cbc32e91/spacy_wrap/layers/clf_transformer_model.py#L61).

**Reply by adrianeboyd** [reply]
Here's a PR to do this in `spacy-transformers` so it's easier to wrap other model types:

https://github.com/explosion/spacy-transformers/pull/332

Here's a sketch of what `spacy-wrap` could like after this PR:

https://github.com/adrianeboyd/spacy-wrap/tree/feature/spacy-transformers-auto

(The `spacy-transformers` requirement would be `>=1.1.7` later if this PR is merged, but this works for testing with PR branch.)

**Reply by KennethEnevoldsen** [reply]
Thanks for this @adrianeboyd! Sorry for getting to it a bit late. It is a wonderful update - will definitely make it easier to maintain and extend spacy-wrap. I have updated . it with the changes.
